### PR TITLE
FlashInfer sampling

### PR DESCRIPTION
**Pull request by tdene**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/arguments.py [hunk 1]
```diff
@@ -1005,6 +1006,16 @@ def validate_args(args, defaults={}):
     ):
         raise ValueError("MXFP8 with inference optimized layers requires FlashInfer >= 0.6.4")
 
+    if args.inference_dynamic_batching_sampling_backend == 'flashinfer':
+        try:
+            import flashinfer  # noqa: F401
+        except ImportError:
+            warnings.warn(
```
Comment (wdykas) [inline megatron/training/arguments.py]: I am of the opinion that we should fail on users on stuff like this.  The logs are already a mess that most people would miss this

File: megatron/core/inference/config.py [hunk 2]
```diff
@@ -320,3 +324,13 @@ def __post_init__(self, verbose: bool):
                 f"prefix_caching_routing_alpha must be in [0, 1], "
                 f"got {self.prefix_caching_routing_alpha}"
             )
+
+        if self.sampling_backend == 'flashinfer':
+            try:
+                import flashinfer  # noqa: F401
+            except ImportError:
+                warnings.warn(
```
Comment (wdykas) [inline megatron/core/inference/config.py]: same thing as below maybe fail here instead of warn?

File: megatron/core/inference/sampling/torch_sampling.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                assert top_k < vocab_size, "top-k is larger than vocab size."
+            modify_logits_for_top_k_filtering(last_token_logits, top_k)
+        elif top_p > 0.0:
+            modify_logits_for_top_p_filtering(last_token_logits, top_p)
+
+        probabilities = last_token_logits.softmax(dim=-1)
+        sampled = torch.multinomial(probabilities, num_samples=1, generator=generator).view(-1)
+
+        if vocab_size:
+            sampled = torch.clamp(sampled, min=0, max=(vocab_size - 1))
```
Comment (wdykas) [inline megatron/core/inference/sampling/torch_sampling.py:91]: What does flash infer do with this? I know this was a topic before for us

File: megatron/core/inference/sampling/flashinfer_sampling.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            scaled = logits[gather_indices[:n], :] / temperature.unsqueeze(1)
+        probs = torch.softmax(scaled, dim=-1)
+
+        # Sentinel values disable filtering:
+        # top_k=vocab_size keeps all tokens, top_p=1.0 keeps the full probability mass.
+        # TODO: Consider changing the disable flags in the `InferenceRequest`.
+        top_k_safe = top_k.masked_fill(top_k == 0, self._vocab_size)
+        top_p_safe = top_p.masked_fill(top_p == 0.0, 1.0)
+        output = torch.empty(n, device=logits.device, dtype=torch.int64)
+        output.copy_(
```
Comment (wdykas) [inline megatron/core/inference/sampling/flashinfer_sampling.py:96]: Just referencing my note from above do we need to do any of the clamping stuff for corner cases of flash infer handles this somehow

**Review by wdykas — APPROVED** [review]
left a few comments but otherwise looks good

File: megatron/core/inference/sampling/torch_sampling.py (hunk 3 above)
Comment (tdene) [reply megatron/core/inference/sampling/torch_sampling.py:91]: This clamp handles a bug that happens in `torch.multinomial` due to FP rounding. It doesn't exist in FlashInfer's kernels.

File: megatron/core/inference/sampling/flashinfer_sampling.py (hunk 4 above)
Comment (tdene) [reply megatron/core/inference/sampling/flashinfer_sampling.py:96]: Yeah, the `_safe` lines convert from our sentinels (0 on both), to FlashInfer-compatible ones.

The FlashInfer kernel is here: https://github.com/flashinfer-ai/flashinfer/blob/main/include/flashinfer/sampling.cuh#L1187

`k >= vocab_size` explicitly short-circuits. `p=1.0` implicitly short-circuits, after entering the kernel's inner GPU loop at least once.

In some tests I ran, the topk topp kernel was 0.8ms, while the non-topk non-topp kernel was 0.6ms. Due to the fact that topp has no explicit short-circuit sentinel.

The decision was made that 0.2ms was not worth recording 2 sets of graphs.

File: megatron/core/inference/config.py (hunk 2 above)
Comment (tdene) [reply megatron/core/inference/config.py]: Done!

File: megatron/training/arguments.py (hunk 1 above)
Comment (tdene) [reply megatron/training/arguments.py]: Done!

File: megatron/core/inference/sampling/flashinfer_sampling.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            flashinfer.sampling.top_k_top_p_sampling_from_probs(
+                probs, top_k_safe, top_p_safe, generator=self._rng
+            )
+        )
```
[the remark this answers is not in the document]
Comment (tdene) [reply megatron/core/inference/sampling/flashinfer_sampling.py:100]: I appreciate your input, Claude, but if you would refer to [the source](https://github.com/flashinfer-ai/flashinfer/blob/v0.5.3/flashinfer/sampling.py) I believe we can show your statement to be incorrect.

`top_k_top_p_sampling_from_probs` returns only samples from 0.5.3 onward.

**Review by santhnm2 — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

File: megatron/core/inference/config.py [hunk 6]
```diff
@@ -297,6 +297,9 @@ class InferenceConfig:
     Defaults to 0, which means no logging.
     """
 
+    sampling_backend: Literal['torch', 'flashinfer'] = 'flashinfer'
```
Comment (ericharper) [inline megatron/core/inference/config.py]: @shanmugamr1992 this change the default and I think the new config will be passed through to nemo rl? Is it okay?

Comment (shanmugamr1992) [reply megatron/core/inference/config.py]: Thats okay, because anyway there is also an MR [#2355](https://github.com/NVIDIA-NeMo/RL/pull/2355/) that @tdene  raised to nemo rl, which he will be testing out on monday (This will be compatible with main)

**Review by shanmugamr1992 — APPROVED** [review]

**Review by ko3n1g — APPROVED** [review]
